lib/service/dataLink.js

```diff
@@ -0,0 +1,214 @@
+  }
+
+  onDeleteByQuery(functionToExecute) {
+    this.register('onDeleteByQuery', functionToExecute);
+  }
+
+  onUpdate(functionToExecute) {
+    this.register('onUpdate', functionToExecute);
+  }
+
+  onGetById(functionToExecute) {
+    this.register('onGetById', functionToExecute);
+  }
+
+  onGetAll(functionToExecute) {
+    this.register('onGetAll', functionToExecute);
+  }
+
+  onGetByQuery(functionToExecute) {
+    this.register('onGetByQuery', functionToExecute);
+  }
+
+  onGetCount(functionToExecute) {
+    this.register('onGetCount', functionToExecute);
+  }
+
+  onGetCountWithQuery(functionToExecute) {
+    this.register('onGetCountWithQuery', functionToExecute);
+  }
+
+  removeHandler(handler) {
+    this.unregister(handler);
+  }
+
+  resolve(dataOp) {
+    if (!this.eventMap[dataOp]) {
+      return notImplementedHandler;
+    }
+
+    return this.eventMap[dataOp];
+  }
+}
+
+exports.serviceObject = serviceObject = (serviceObjectName) => {
+  if (!registeredServiceObjects[serviceObjectName]) {
+    registeredServiceObjects[serviceObjectName] = new ServiceObject(serviceObjectName);
+  }
+
+  return registeredServiceObjects[serviceObjectName];
+};
+
+exports.process = (task, modules, callback) => {
+  if (task.request.serviceObjectName === null) {
+    const result = task.response;
+    result.body = kinveyErrors.generateKinveyError('NotFound', 'ServiceObject name not found');
+    result.statusCode = result.body.statusCode;
+    delete result.body.statusCode;
+    return callback(task);
+  }
+
+  // Handle a bug in KCS - if query is part of the task, put it in the request
+  if (!task.request.query) {
+    task.request.query = task.query;
+  }
+
+  const serviceObjectToProcess = serviceObject(task.request.serviceObjectName);
+  let dataOp = '';
+  const dataLinkCompletionHandler = kinveyCompletionHandler(task, callback);
+
+  try {
+    task.request.body = JSON.parse(task.request.body);
+  } catch (e) {
+    if (task.request.body && typeof task.request.body !== 'object') {
+      const result = task.response;
+      result.body = kinveyErrors.generateKinveyError('BadRequest', 'Request body is not JSON');
+      result.statusCode = result.body.statusCode;
+      delete result.body.statusCode;
+      return callback(task);
+    }
+  }
+
+  try {
+    task.request.query = JSON.parse(task.request.query);
+  } catch (error) {
+    if (task.request.query && typeof task.request.query !== 'object') {
+      const result = task.response;
+      result.body = kinveyErrors.generateKinveyError('BadRequest', 'Request query is not JSON');
+      result.statusCode = result.body.statusCode;
+      delete result.body.statusCode;
+      return callback(task);
+    }
+  }
+
+  if (task.method === 'POST') {
+    dataOp = 'onInsert';
+  } else if (task.method === 'PUT') {
+    dataOp = 'onUpdate';
+  } else if (task.method === 'GET' && task.endpoint !== '_count') {
+    const taskRequest = _.get(task, 'request', {});
```

Comment:
Any reason to not just do `const taskRequest = task.request || {}`?

(or, if you want to avoid checking truthiness -- `const taskRequest = task.hasOwnProperty('request') ? task.request : {}`)